Using the connector in spark-shell (Spark 2.4.3)

It is currently not possible to use the connector in spark-shell as it blows up with NCDF. Please investigate and offer a way to make that possible.

Don't forget to update the `README.md`.

```
$ spark-shell --jars target/scala-2.11/splice-machine-spark-connector_2.11-0.1.jar
...
scala> spark.version
res0: String = 2.4.3

val user = "splice"
val password = "admin"
val url = s"jdbc:splice://localhost:1527/splicedb;user=$user;password=$password"
val table = "t1"
val t1 = spark.read.format("spliceV1").option("url", url).option("table", table).load
java.lang.NoClassDefFoundError: com/splicemachine/spark/splicemachine/SplicemachineContext
  at splice.v1.SpliceDataSourceV1.createRelation(SpliceDataSourceV1.scala:25)
  at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:318)
  at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:223)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:211)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:167)
  ... 49 elided
Caused by: java.lang.ClassNotFoundException: com.splicemachine.spark.splicemachine.SplicemachineContext
  at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  ... 54 more
```

Having a tough fight with maven classifier "machinery" and spark-shell.

splice uses maven classifier for Hadoop distro, e.g. `cdh5.14.0`. spark-shell's `--packages` does not seem to support classifiers (https://issues.apache.org/jira/browse/SPARK-20075). The idea was to include classifier-based deps in `--jars` while the others using `--packages` to cut the necessary deps on the command line.

The strategy does not seem to work though.

In summary:

- hbase_storage uses ZeroCopyLiteralByteString from splice_protocol that extends LiteralByteString from protobuf-java

- so using --jars with hbase_storage and --packages for the others won't work because:

```
Caused by: java.lang.IllegalAccessError: class com.google.protobuf.ZeroCopyLiteralByteString cannot access its superclass com.google.protobuf.LiteralByteString
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  at java.security.AccessController.doPrivileged(Native Method)
  at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  at com.splicemachine.si.impl.CoprocessorTxnStore.recordNewTransaction(CoprocessorTxnStore.java:122)
```

[sbt-assembly](https://github.com/sbt/sbt-assembly) plugin configured to build the connector (as an uber-jar). The changes are in the [sbt-assembly branch](../tree/sbt-assembly).

When executed in `spark-shell` fails with the same `IllegalAccessError` as before (with `--jars` and `--packages`).

```
$ spark-shell --jars target/scala-2.11/splice-machine-spark-connector-assembly-0.1.jar
...
Caused by: java.lang.IllegalAccessError: class com.google.protobuf.ZeroCopyLiteralByteString cannot access its superclass com.google.protobuf.LiteralByteString
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  at java.security.AccessController.doPrivileged(Native Method)
  at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  at com.splicemachine.si.impl.CoprocessorTxnStore.recordNewTransaction(CoprocessorTxnStore.java:122)
  at com.splicemachine.si.impl.ClientTxnLifecycleManager.elevateTransaction(ClientTxnLifecycleManager.java:195)
  at com.splicemachine.si.impl.txn.ReadOnlyTxn.elevateToWritable(ReadOnlyTxn.java:223)
  at com.splicemachine.si.impl.TransactionImpl.doElevate(TransactionImpl.java:357)
  at com.splicemachine.si.impl.TransactionImpl.elevate(TransactionImpl.java:293)
  at com.splicemachine.derby.impl.store.access.SpliceTransaction.elevate(SpliceTransaction.java:114)
  at com.splicemachine.derby.impl.store.access.SpliceTransactionManager.elevate(SpliceTransactionManager.java:1263)
  at com.splicemachine.derby.impl.sql.catalog.SpliceDataDictionary.updateSystemSchemasView(SpliceDataDictionary.java:1371)
  at com.splicemachine.db.impl.sql.catalog.DataDictionaryImpl.boot(DataDictionaryImpl.java:536)
  at com.splicemachine.derby.impl.sql.catalog.SpliceDataDictionary.boot(SpliceDataDictionary.java:476)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.boot(BaseMonitor.java:1963)
  at com.splicemachine.db.impl.services.monitor.TopService.bootModule(TopService.java:337)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.startModule(BaseMonitor.java:546)
  at com.splicemachine.db.impl.services.monitor.FileMonitor.startModule(FileMonitor.java:57)
  at com.splicemachine.db.iapi.services.monitor.Monitor.bootServiceModule(Monitor.java:435)
  at com.splicemachine.db.impl.db.BasicDatabase.boot(BasicDatabase.java:189)
  at com.splicemachine.derby.impl.db.SpliceDatabase.boot(SpliceDatabase.java:120)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.boot(BaseMonitor.java:1963)
  at com.splicemachine.db.impl.services.monitor.TopService.bootModule(TopService.java:337)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.bootService(BaseMonitor.java:1798)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.startProviderService(BaseMonitor.java:1665)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.findProviderAndStartService(BaseMonitor.java:1543)
  at com.splicemachine.db.impl.services.monitor.BaseMonitor.startPersistentService(BaseMonitor.java:967)
  at com.splicemachine.db.iapi.services.monitor.Monitor.startPersistentService(Monitor.java:557)
  at com.splicemachine.db.impl.jdbc.EmbedConnection.bootDatabase(EmbedConnection.java:2278)
  ... 66 more
```

--- a/README.md
+++ b/README.md
@@ -115,10 +115,25 @@ The below session uses `spark-shell` for demonstration purposes.
 **NOTE**: Start Splice Machine, e.g. `./start-splice-cluster -p cdh5.14.0 -bl`.
 
 ```
-$ spark-shell --jars target/scala-2.11/splice-machine-spark-connector-assembly-0.1.jar
-...
-scala> spark.version
-res0: String = 2.4.3
+// Create table first
+$ ./sqlshell.sh
+
+splice> create table t1 (id int, name varchar(50));
+0 rows inserted/updated/deleted
+
+splice> insert into t1 values (0, 'The connector works!');
+1 row inserted/updated/deleted
+
+// Use Spark 2.2 with Hadoop 2.6 or compatible version
+// e.g. spark-2.2.3-bin-hadoop2.6.tgz
+// https://archive.apache.org/dist/spark/spark-2.2.3/
+// java.lang.NoSuchMethodError: org.apache.hadoop.mapreduce.InputSplit.getLocationInfo()[Lorg/apache/hadoop/mapred/SplitLocationInfo;
+
+$ spark-shell --driver-class-path target/scala-2.11/splice-machine-spark-connector-assembly-0.1.jar
+
+assert(spark.version == "2.4.3", "The connector works just fine with Spark 2.4.3")
+
+assert(spark.version != "2.2.3", "FIXME: The connector does not work with Spark 2.2.3") 
 
 val user = "splice"
 val password = "admin"
@@ -127,9 +142,9 @@ val table = "t1"
 val t1 = spark.read.format("spliceV1").option("url", url).option("table", table).load
 
 scala> t1.show
-+---+--------------+
-| id|          name|
-+---+--------------+
-|  0|splice machine|
-+---+--------------+
++---+--------------------+
+| ID|                NAME|
++---+--------------------+
+|  0|The connector works!|
++---+--------------------+
 ```

--- a/README.md
+++ b/README.md
@@ -115,7 +115,7 @@ The below session uses `spark-shell` for demonstration purposes.
 **NOTE**: Start Splice Machine, e.g. `./start-splice-cluster -p cdh5.14.0 -bl`.
 
 ```
-$ spark-shell --jars target/scala-2.11/splice-machine-spark-connector_2.11-0.1.jar
+$ spark-shell --jars target/scala-2.11/splice-machine-spark-connector-assembly-0.1.jar
 ...
 scala> spark.version
 res0: String = 2.4.3

--- a/build-assembly.sbt
+++ b/build-assembly.sbt
@@ -1,0 +1,88 @@
+// Don't execute tests
+test in assembly := {}
+
+// Exclude duplicates
+
+// NOTE excludeAll in build.sbt won't work
+// Use assemblyMergeStrategy and assemblyExcludedJars
+// * https://github.com/sbt/sbt-assembly/issues/35
+// * https://stackoverflow.com/q/41894055/1305344
+
+assemblyMergeStrategy in assembly := {
+  // FIXME How do others deal with UnusedStubClass.class in so many Spark deps?
+  case PathList("org", "apache", "spark", "unused", "UnusedStubClass.class") => MergeStrategy.first
+  // FIXME Why do the two come with the same package-info.class?
+  // hadoop-yarn-api-2.6.0-cdh5.14.0.jar
+  // hadoop-yarn-common-2.6.0-cdh5.14.0.jar
+  case PathList("org", "apache", "hadoop", "yarn", ps @ _*) if ps.lastOption.exists(_ endsWith "package-info.class") => MergeStrategy.first
+  // netty-all-4.0.43.Final.jar VS hadoop-hdfs-2.6.0-cdh5.14.0.jar
+  case PathList("META-INF", "io.netty.versions.properties")  => MergeStrategy.first
+  // db-engine-2.8.0.1926.jar VS splice_encoding-2.8.0.1926.jar
+  case PathList("com", "splicemachine", "utils", _*)  => MergeStrategy.first
+  case PathList(ps @ _*) if ps.last endsWith ".css"  => MergeStrategy.first
+  case x =>
+    val oldStrategy = (assemblyMergeStrategy in assembly).value
+    oldStrategy(x)
+}
+
+assemblyExcludedJars in assembly := {
+  val excludedDeps = Set(
+    // Conflicts with splice_aws-2.8.0.1926-shade.jar
+    "aws-java-sdk-bundle-1.11.134.jar",
+    "aws-java-sdk-kms-1.11.82.jar",
+    "ion-java-1.0.1.jar",
+    // Conflicts with hadoop-core-2.6.0-mr1-cdh5.14.0.jar
+    "hadoop-mapreduce-client-core-2.6.0-cdh5.14.0.jar",
+    "hadoop-mapreduce-client-common-2.6.0-cdh5.12.0.jar",
+    // Conflicts with Spark 2.2.0.cloudera2
+    // FIXME What dep brings Spark 2.1.0?
+    "spark-yarn_2.11-2.1.0.jar",
+    // Conflicts with commons-beanutils-1.7.0.jar
+    // http://commons.apache.org/proper/commons-beanutils/
+    // Bean Collections has an additional dependency on Commons Collections.
+    "commons-collections-3.2.2.jar",
+    // Conflicts with commons-beanutils-core-1.8.0.jar
+    // NOTE a dependency version mismatch
+    "commons-beanutils-1.7.0.jar",
+    // Included in / Conflicts with
+    // jython-standalone-2.5.3.jar
+    "jline-0.9.94.jar",
+    "xercesImpl-2.9.1.jar",
+    // Conflicts with jersey-core-1.9.jar
+    "javax.ws.rs-api-2.0.1.jar",
+    // Included in / Conflicts with
+    // jcl-over-slf4j-1.7.5.jar
+    "commons-logging-1.2.jar",
+    // Included in / Conflicts with
+    // aopalliance-repackaged-2.4.0-b34.jar
+    "aopalliance-1.0.jar",
+    // Included in / Conflicts with
+    // servlet-api-2.5-6.1.14.jar
+    "servlet-api-2.5-20081211.jar",
+    // Included in / Conflicts with
+    // servlet-api-2.5-6.1.14.jar from Jetty
+    "servlet-api-2.5.jar",
+    "jsp-api-2.1-6.1.14.jar",
+    "jsp-api-2.1.jar",
+    // Included in / Conflicts with
+    // hbase_sql-2.8.0.1926-cdh5.14.0.jar
+    // jython-standalone-2.5.3.jar
+    "guava-20.0.jar",
+    // Included in / Conflicts with
+    // jersey-server-2.22.2.jar
+    // NOTE a dependency version mismatch
+    "jersey-server-1.9.jar",
+    // Included in / Conflicts with
+    // servlet-api-2.5-6.1.14.jar from Jetty
+    // NOTE a dependency version mismatch
+    "javax.servlet-api-3.1.0.jar",
+    // Included in / Conflicts with
+    // javax.inject-2.4.0-b34.jar from Glassfish
+    "javax.inject-1.jar",
+    // Others
+    "jasper-runtime-5.5.23.jar",
+    "findbugs-annotations-1.3.9-1.jar"
+  )
+  val cp = (fullClasspath in assembly).value
+  cp filter { f => excludedDeps.contains(f.data.getName) }
+}

--- a/build-assembly.sbt
+++ b/build-assembly.sbt
@@ -25,6 +25,7 @@ assemblyMergeStrategy in assembly := {
     oldStrategy(x)
 }
 
+// FIXME There has to be a better way to exclude all the jars (!)
 assemblyExcludedJars in assembly := {
   val excludedDeps = Set(
     // Conflicts with splice_aws-2.8.0.1926-shade.jar
@@ -79,10 +80,70 @@ assemblyExcludedJars in assembly := {
     // Included in / Conflicts with
     // javax.inject-2.4.0-b34.jar from Glassfish
     "javax.inject-1.jar",
+    // FIXME It is assumed to exist earlier. Make sure nothing got broken
+    // Have to be excluded to let shading in sbt-assembly work
+    "jython-standalone-2.5.3.jar",
+    "netty-3.10.6.Final.jar",
+    // Exclude Spark jars
+    "spark-launcher_2.11-2.2.0.cloudera2.jar",
+    "spark-yarn_2.11-2.1.0.jar",
+    "spark-network-common_2.11-2.2.0.cloudera2.jar",
+    "spark-network-shuffle_2.11-2.2.0.cloudera2.jar",
+    "spark-unsafe_2.11-2.2.0.cloudera2.jar",
+    "spark-tags_2.11-2.2.0.cloudera2.jar",
+    "spark-core_2.11-2.2.0.cloudera2.jar",
+    // FIXME Should spark-avro be excluded as well?
+    "spark-avro_2.11-4.0.0.jar",
+    // Hadoop
+    "hadoop-mapreduce-client-jobclient-2.6.0-cdh5.12.0.jar",
+    "hadoop-aws-2.6.0-cdh5.12.0.jar",
+    "hadoop-yarn-server-common-2.6.0-cdh5.12.0.jar",
+    "hadoop-hdfs-2.6.0-cdh5.14.0.jar",
+    "hadoop-hdfs-2.6.0-cdh5.14.0-tests.jar",
+    "hadoop-core-2.6.0-mr1-cdh5.14.0.jar",
+    "hadoop-client-2.6.0-cdh5.12.0.jar",
+    "hadoop-mapreduce-client-app-2.6.0-cdh5.12.0.jar",
+    "hadoop-mapreduce-client-common-2.6.0-cdh5.12.0.jar",
+    "hadoop-auth-2.6.0-cdh5.14.0.jar",
+    "hadoop-mapreduce-client-shuffle-2.6.0-cdh5.12.0.jar",
+    "hadoop-yarn-common-2.6.0-cdh5.14.0.jar",
+    "hadoop-yarn-server-nodemanager-2.6.0-cdh5.12.0.jar",
+    "hadoop-mapreduce-client-core-2.6.0-cdh5.14.0.jar",
+    "hadoop-yarn-server-web-proxy-2.2.0.jar",
+    "hbase-hadoop2-compat-1.2.0-cdh5.14.0.jar",
+    "hadoop-yarn-client-2.6.0-cdh5.12.0.jar",
+    "hadoop-annotations-2.6.0-cdh5.14.0.jar",
+    "hadoop-yarn-api-2.6.0-cdh5.14.0.jar",
+    "hadoop-common-2.6.0-cdh5.14.0.jar",
     // Others
     "jasper-runtime-5.5.23.jar",
-    "findbugs-annotations-1.3.9-1.jar"
+    "findbugs-annotations-1.3.9-1.jar",
+    "scalactic_2.11-3.0.8.jar",
+    "scala-reflect-2.11.12.jar",
+    "scala-xml_2.11-1.2.0.jar",
+    "scala-parser-combinators_2.11-1.0.4.jar",
+    "scala-compiler-2.11.12.jar",
+    // Already in Spark
+    "lz4-1.3.0.jar"
   )
   val cp = (fullClasspath in assembly).value
-  cp filter { f => excludedDeps.contains(f.data.getName) }
+  cp filter { f =>
+    val r = excludedDeps.contains(f.data.getName)
+    val name = f.data.getName
+    if (name.startsWith("protobuf-java") ||
+//        name.startsWith("jython-standalone") ||
+//        name.startsWith("netty") ||
+        name.startsWith("spark")) {
+      println(s">>> Found: $name")
+    }
+    r
+  }
 }
+
+// Shading protobuf-java-2.5.0.jar as it is part of Spark (and spark-shell)
+// $SPARK_HOME/jars/protobuf-java-2.5.0.jar
+// TIP Use getClass().getProtectionDomain().getCodeSource() to know the location of a class
+assemblyShadeRules in assembly := Seq(
+  ShadeRule.rename("com.google.protobuf.**" -> "shaded.protobuf.@1").inAll,
+  ShadeRule.rename("io.netty.**" -> "shaded.netty.@1").inAll
+)

--- a/build.sbt
+++ b/build.sbt
@@ -17,6 +17,11 @@ val hadoopVersion = s"2.6.0-$envClassifier"
 // java.lang.NoClassDefFoundError: com/splicemachine/access/HConfiguration
 // https://stackoverflow.com/a/46763742/1305344
 
+val excludedDeps = Seq(
+  ExclusionRule(organization = "org.xerial.snappy", name = "snappy-java"),
+  ExclusionRule(organization = "tomcat", name = "jasper-compiler")
+)
+
 libraryDependencies ++= Seq(
   "splice_spark",
   "hbase_sql",
@@ -32,11 +37,12 @@ val scalaUtilClassifier = Def.setting {
 }
 libraryDependencies += spliceDep("scala_util", scalaUtilClassifier.value)
 
-libraryDependencies += "org.apache.hadoop" % "hadoop-common" % hadoopVersion
+libraryDependencies += "org.apache.hadoop" % "hadoop-common" % hadoopVersion excludeAll(excludedDeps: _*)
+
 // Required to ensure proper dependency
 // (otherwise cdh5.12.0 version was resolved and used)
-libraryDependencies += "org.apache.hadoop" % "hadoop-mapreduce-client-core" % hadoopVersion
-libraryDependencies += "org.apache.hbase" % "hbase-server" % hbaseVersion
+libraryDependencies += "org.apache.hadoop" % "hadoop-mapreduce-client-core" % hadoopVersion excludeAll (excludedDeps: _*)
+libraryDependencies += "org.apache.hbase" % "hbase-server" % hbaseVersion excludeAll (excludedDeps: _*)
 
 resolvers +=
   "splicemachine-public" at "http://repository.splicemachine.com/nexus/content/groups/public"
@@ -46,7 +52,7 @@ resolvers +=
 resolvers += Resolver.mavenLocal
 
 // com.fasterxml.jackson.databind.JsonMappingException: Incompatible Jackson version: 2.9.2
-libraryDependencies += "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.9.9" force()
+libraryDependencies += "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.9.9" force() excludeAll (excludedDeps: _*)
 
 updateOptions := updateOptions.value.withLatestSnapshots(false)
 
@@ -63,6 +69,6 @@ libraryDependencies += "org.scalactic" %% "scalactic" % scalatestVer
 libraryDependencies += "org.scalatest" %% "scalatest" % scalatestVer % Test
 parallelExecution in Test := false
 
-def spliceDep(name: String, classifier: String = envClassifier): ModuleID = {
-  "com.splicemachine" % name % spliceVersion classifier classifier withSources()
+def spliceDep(name: String, classfr: String = envClassifier): ModuleID = {
+  "com.splicemachine" % name % spliceVersion classifier classfr withSources() excludeAll (excludedDeps: _*)
 }

--- a/project/assembly.sbt
+++ b/project/assembly.sbt
@@ -1,0 +1,1 @@
+addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.14.10")

